### ISSUE TITLE
[CRITICAL] Backwards Compatibility - Moving expenses between reports

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#33177]
### Problem 

As part of the workspace chats for [paid group policies](https://github.com/Expensify/Expensify/issues/309695), we need to make sure actions possible to take in OldDot are correctly reflected in NewDot.

One of them is moving expenses between reports, this influences how the expense reports are shown in newdot.

The high level details are explained in the Design doc section [here](https://docs.google.com/document/d/1arowjiiYSu4fyu0CYrB8748UOz4oEDaWZzmXpGKwg-k/edit#bookmark=id.hnqnayp68r2g) in terms of what we need to do.

## Solution
In OldDot, users can move an expense/money request from one report to another report. With the App IOU report actions approach, this however means we need to handle some additional stuff when such a move occurs.

When an expense is moved from ReportA to ReportB and both reports are on the same policy-expense-chat-enabled policies, we need to in UpdateTransaction:
- [x] Update the reportID in the transactions table (this is already happening)
- [x] Change the reportID of the IOU report action linked to the transaction to the new reportID
    - [x] The transaction thread does not have parent report ID NVP, only the parent report action which was moved to the new report so no need to change anything here
- [x] Recompute the report cache to get updated total
- [x] In case the report we are moving to has been created after the expense/ IOU report action has been created, we need to update the created property of the report and the created timestamp of the CREATED report action of the expense report to be before the transaction created timestamp
    - [x] We can follow what we already do in PayMoneyRequest [here](https://github.com/Expensify/Auth/blob/757e2320dca05cc39473d48285901b6ace934425/auth/command/PayMoneyRequest.cpp#L313-L343)

---

- [ ] If we move the expense to a report on policy without workspace chat, we will delete the policy expense chat aspects of the expense report. We will delete the report preview, the parent reportID rNVP, and parent report action rNVP.

- [x] If we unreport an expense, let's move the IOU report action to the deleted report so it's not shown on any expense report.

Please bring any questions to the wave6 room, the design doc is not written in a lot of detail but the things we need to do are captured there in plain English

**Comment by yuwenmemon** [comment]
Looking into this now.

**Comment by yuwenmemon** [comment]
> When an expense is moved from ReportA to ReportB and both reports are on the same policy-expense-chat-enabled policies, we need to in UpdateTransaction

I think we mean [**ReportTransactions** ](https://github.com/Expensify/Auth/blob/717fafc56903e0bda87ea6e7761a45f131d10557/auth/command/ReportTransactions.cpp)here? Or, am I missing something?

**Comment by yuwenmemon** [comment]
Thinking of how to break this up into as small pieces as possible. Going to tackle this first:?

> If we move the expense to a report on policy without workspace chat, we will delete the policy expense chat aspects of the expense report. We will delete the report preview, the parent reportID rNVP, and parent report action rNVP.

**Comment by yuwenmemon** [comment]
Asked a question about the above here: https://expensify.slack.com/archives/C02MW39LT9N/p1704494359050069

**Comment by yuwenmemon** [comment]
Asked another [question](https://expensify.slack.com/archives/C02MW39LT9N/p1704496054840879) about this:

> In case the report we are moving to has been created after the expense/ IOU report action has been created, we need to update the created property of the report and the created timestamp of the CREATED report action of the expense report to be before the transaction created timestamp

**Comment by yuwenmemon** [comment]
Okay trying to summarize what this looks like presently so I can understand what we need to fix. I currently create a Money Request in NewDot. This creates the following report `115475633052248`:
![Screenshot 2024-01-05 at 3 34 43 PM](https://github.com/Expensify/App/assets/4741899/3c8d957f-e6a5-458e-9cf3-8e189ab9bcdf)
![Screenshot 2024-01-05 at 3 35 14 PM](https://github.com/Expensify/App/assets/4741899/5df2254e-c1d3-44ae-8417-969388d5efd3)
```
sqlite> select * from reports where reportID=115475633052248;
   created = 2024-01-05 23:28:50
 submitted = 2024-01-05 23:28:50
  approved = 
  reportID = 115475633052248
 accountID = 32
 managerID = 2
reportName = Expense Report #115475633052248
     state = 0
     total = -500
  currency = USD
cachedData = {"submitted":"","submittedTo":"","approved":"","unapproved":"","reimbursed":"","shared":"","type":"expense","retracted":"","expensify_policyID":"D5885A35873BF696","policyLastModified":1704406113036338,"isSmartReport":false,"cacheComputedBy":32,"inbox":{"violations":{"oldestExpenseDate":null,"newestExpenseDate":null,"maxAge":null,"potentialRTERViolations":[],"cacheableViolationTypes":["violation"],"ignoredWarnings":[],"hasTransactionsWithDifferentCurrency":false,"violationsReportAmount":-500,"totalViolations":1},"policyType":"corporate","firstSubmittedDate":"2024-01-05 23:28:50","wasComputedInReportCache":true},"reimburserEmail":"yuwen@expensify.com","queuedForAutoExport":{"queued":false,"connections":[],"exportFailed":false,"failedConnections":[]},"hasBeenExportedToIntegration":false,"wasReimbursedViaACH":false,"statementCardID":null,"statementBank":null,"statementPeriodEndDay":null,"isWaitingOnBankAccount":false,"isABAReimbursed":false,"expectedReimbursementDate":null,"canMarkReimbursementReceived":true,"hasReimbursableTotal":true,"hasTransactions":true,"hasBillableTransactions":false,"reportOwners":[],"managerOnVacation":"","isSmartReportOnFile":false,"invoiceID":0,"billID":0}
    status = 0
```

```
sqlite> select * from reportNameValuePairs where reportID=115475633052248;
reportID         name                     value                                     
---------------  -----------------------  ------------------------------------------
115475633052248  expensify_policyID       D5885A35873BF696                          
115475633052248  lastReadTime_32          2024-01-05 23:28:50.407                   
115475633052248  notificationPreferences  {"2":"hidden","32":"hidden","46":"hidden"}
115475633052248  parentReportActionID     7370303887022919                          
115475633052248  parentReportID           7352862044010264                          
115475633052248  type                     expense       
```

The transaction `2589566964639939880` is on this report:
```
sqlite> select * from transactions where reportID = 115475633052248;
created     transactionID        cardID  reportID         receiptID  amount  merchant  mcc  state  comment              tag  externalID  memo  inserted             currency  modifiedCreated  modifiedAmount  modifiedMerchant  modifiedMCC  category  transactionHash                               modifiedCurrency
----------  -------------------  ------  ---------------  ---------  ------  --------  ---  -----  -------------------  ---  ----------  ----  -------------------  --------  ---------------  --------------  ----------------  -----------  --------  --------------------------------------------  ----------------
2024-01-05  2589566964639939880  31      115475633052248             -500    FIVERRRR       -3     {"comment":"Fiver"}                         2024-01-05 23:28:50  USD                                                                                 hash6798A0234119C3CFA6D3FB7F6C73E59E9C5264B3                  
sqlite> 
```
And it has an IOU report action that corresponds to it:
```
sqlite> select * from reportActions where reportActionID=8969807020746277047;
created                  reportID         accountID  action  message                                                       reportActionID     
-----------------------  ---------------  ---------  ------  ------------------------------------------------------------  -------------------
2024-01-05 23:28:50.406  115475633052248  32         IOU     {"IOUReportID":115475633052248,"IOUTransactionID":"258956696  8969807020746277047
                                                             4639939880","amount":500,"comment":"Fiver","currency":"USD",                     
                                                             "lastModified":"2024-01-05 23:28:50.406","participantAccount                     
                                                             IDs":[32,0],"type":"create"}    
```

That Money Request has a transaction thread `4028651028712853`:
![Screenshot 2024-01-05 at 3 36 45 PM](https://github.com/Expensify/App/assets/4741899/c6b1d472-01d9-491b-8ce2-408ea4db2e7d)
```
sqlite> select * from reports where reportID=4028651028712853;
   created = 2024-01-05 23:28:50
 submitted = 
  approved = 
  reportID = 4028651028712853
 accountID = 0
 managerID = 
reportName = Chat Report
     state = 0
     total = 0
  currency = USD
cachedData = 
    status = 0
```
```
sqlite> select * from reportNameValuePairs where reportID=4028651028712853;
reportID          name                     value                                     
----------------  -----------------------  ------------------------------------------
4028651028712853  expensify_policyID       D5885A35873BF696                          
4028651028712853  lastReadTime_32          2024-01-05 23:36:08.261                   
4028651028712853  notificationPreferences  {"2":"hidden","32":"hidden","46":"hidden"}
4028651028712853  parentReportActionID     8969807020746277047                       
4028651028712853  type                     chat     
```

### I now change the transaction to go on another report in the same policy.

New reportID: `8616140694773666`

If I click on the IOU report action, it still takes me to `115475633052248`:
![Screenshot 2024-01-05 at 4 14 18 PM](https://github.com/Expensify/App/assets/4741899/38ff09dd-bf53-4dfd-9d29-b8bf0ef95feb)

Makes sense because the IOU report action is unchanged:
```
sqlite> select * from reportActions where reportActionID=8969807020746277047;
created                  reportID         accountID  action  message                                                       reportActionID     
-----------------------  ---------------  ---------  ------  ------------------------------------------------------------  -------------------
2024-01-05 23:28:50.406  115475633052248  32         IOU     {"IOUReportID":115475633052248,"IOUTransactionID":"258956696  8969807020746277047
                                                             4639939880","amount":500,"comment":"Fiver","currency":"USD",                     
                                                             "lastModified":"2024-01-05 23:28:50.406","participantAccount                     
                                                             IDs":[32,0],"type":"create"}  
```

So this makes sense to me:
> Change the reportID of the IOU report action linked to the transaction to the new reportID

We need to update the `IOUReportID` in the JSON above to be the new reportID.

Looking at the old report:
```
   created = 2024-01-05 23:28:50
 submitted = 2024-01-05 23:28:50
  approved = 
  reportID = 115475633052248
 accountID = 32
 managerID = 2
reportName = Expense Report #115475633052248
     state = 0
     total = 0
  currency = USD
cachedData = 
    status = 0
```

... All good with the report object itself in the db ✅ 

```
sqlite> select * from reportNameValuePairs where reportID=115475633052248;
reportID         name                     value                                     
---------------  -----------------------  ------------------------------------------
115475633052248  expensify_policyID       D5885A35873BF696                          
115475633052248  lastReadTime_32          2024-01-06 00:14:19.052                   
115475633052248  notificationPreferences  {"2":"hidden","32":"hidden","46":"hidden"}
115475633052248  parentReportActionID     7370303887022919                          
115475633052248  parentReportID           7352862044010264                          
115475633052248  type                     expense  
```

... rNVPs for the old report are unchanged

The transaction itself gets the reportID changed to the new one:
```
sqlite> select * from transactions where transactionID=2589566964639939880;
created     transactionID        cardID  reportID          receiptID  amount  merchant  mcc  state  comment              tag  externalID  memo  inserted             currency  modifiedCreated  modifiedAmount  modifiedMerchant  modifiedMCC  category       transactionHash                               modifiedCurrency
----------  -------------------  ------  ----------------  ---------  ------  --------  ---  -----  -------------------  ---  ----------  ----  -------------------  --------  ---------------  --------------  ----------------  -----------  -------------  --------------------------------------------  ----------------
2024-01-05  2589566964639939880  31      8616140694773666             -500    FIVERRRR       3      {"comment":"Fiver"}                         2024-01-05 23:28:50  USD                                                                       Uncategorized  hash6798A0234119C3CFA6D3FB7F6C73E59E9C5264B3  USD    
``` 

And then looking at the new report there is no cachedData, as with the old report:
```
sqlite> select * from reports where reportID="8616140694773666";
   created = 2024-01-06 00:12:08
 submitted = 
  approved = 
  reportID = 8616140694773666
 accountID = 32
 managerID = 
reportName = New Report To Transfer Expenses To
     state = 0
     total = -500
  currency = USD
cachedData = 
    status = 0
```

**Comment by yuwenmemon** [comment]
Okay so will tackle this one first:
> Change the reportID of the IOU report action linked to the transaction to the new reportID

**Comment by yuwenmemon** [comment]
Have a WIP PR up, just to make sure I'm on the right track for tackling the tight scope of "Change the reportID of the IOU report action linked to the transaction to the new reportID"

**Comment by yuwenmemon** [comment]
Looking at:

>  Recompute the report cache to get updated total

I _believe_ we already do this here:

https://github.com/Expensify/Auth/blob/63f66f1a25f96329e719ad13a1a4739c531d2be3/auth/lib/Report.cpp#L2405-L2418

So I _think_ we may be ready do have the first PR do:

> Change the reportID of the IOU report action linked to the transaction to the new reportID

and then do

>  If we move the expense to a report on policy without workspace chat, we will delete the policy expense chat aspects of the expense report. We will delete the report preview, the parent reportID rNVP, and parent report action rNVP.

and 

>  If we unreport an expense, let's move the IOU report action to the deleted report so it's not shown on any expense report.

In separate follow-up PRs - what do you think of this plan @mountiny? If :+1: I'll add tests to the above WIP PR and get it ready for review.

**Comment by mountiny** [comment]
I love this plan, thanks for digging into this one!

**Comment by yuwenmemon** [comment]
PR is approved for this! @flodnv had a question but I think we answered it! ....Should I just merge?

**Comment by yuwenmemon** [comment]
Closing this in favor of https://github.com/Expensify/Expensify/issues/364642

Moving expenses _between_ reports has been deployed 🎉
